pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go

```diff
@@ -235,32 +250,87 @@ func nodesWherePreemptionMightHelp(nodes []*framework.NodeInfo, m framework.Node
 	return potentialNodes
 }
 
+// calculateNumCandidates returns the number of candidates the findCandidates
+// method must produce from dry running based on the constraints given by
+// <maxCandidateNodesPercentage> and <minCandidates>. The number of candidates
+// returned will never be greater than <numNodes>.
+func (pl *DefaultPreemption) calculateNumCandidates(numNodes int32) int32 {
+	n := (numNodes * pl.args.MaxCandidateNodesPercentage) / 100
+	if n < pl.args.MinCandidates {
+		n = pl.args.MinCandidates
+	}
+	if n > numNodes {
+		n = numNodes
+	}
+	return n
+}
+
+// calculateOffset returns an <offset> and <numCandidates> that the caller must
+// use to find candidates. A variable offset is used as the starting index when
+// dry running preemption so that the same set of nodes don't get preempted
+// every time. If <offset> + <numCandidates> runs past the number of nodes, the
+// caller is expected to wrap around the node list.
+func (pl *DefaultPreemption) calculateOffset(numNodes int32) (int32, int32) {
+	offset := pl.candidateOffset
+	numCandidates := pl.calculateNumCandidates(numNodes)
+	if pl.candidateOffset >= numNodes {
```

Comment:
We aslo need update `offset` if `offset >= numNodes`?

Reply:
not a big deal, as we use mod later, but yeah, we can

Reply:
Done.